RFC/RFC-14-p2p-adapter.md

```diff
@@ -0,0 +1,348 @@
+---
+layout: doc
+rfc: 14
+date: 2022-10-17
+title: P2P adapter with routing service
+status: DRAFT
+authors:
+  - hugoArregui
+  - agusAldasoro
+---
+
+## Abstract
+
+This document describes a possible implementation for a P2P adapter using a fallback messaging service. For more information about what an adapter is, check [ADR-81](/ADR/ADR-81-minimum-comms-transport.md).
+
+## Need
+
+A P2P adapter is needed in order to provide a light mechanism for comms (compared to [RFC-5: comms service using websocket](/RFC/RFC-5-ws-comms.md)). This adapter should guarantee the delivery of a message to any peer in the network (or cluster, [real-time clustering of users in Islands (ADR-35)](/adr/ADR-35)).
```

Comment:
What real-world problem is this RFC trying to address? It seems like a P2P+TURN server mesh should be compliant with all the things mentioned here, and a way less complex solution to the described problem. Or maybe I don't yet see the key point of this proposal

Reply:
> I think that adding a MessageService defeats the purpose of a lightweight protocol, instead I'd recommend to rely on TURN and P2P mesh.

Well, the message service is a fallback for when cluster happens in the network, the idea of this proposal is to have a algorithm that's simple, have a lot of visibility in what's happening in the network and is subject to improvements. Selecting random candidates may end up reducing the probably of forming internal clusters and thus the message service won't be used. But we have a guarantee that even if clusters appear, we will deal with the situation. Furthermore, if we have clusters and we decide that's a problem, we can suggest new ideas (probably different implementations on top of this RFC) for the server to suggest connections in order to bridge and unify cluster or any others. 

> 
> For the routing service, it seems like we are proposing to re-implement BGP, I'd recommend to take a look at that protocol it is the protocol that literally keeps the world running.

I think we are probably implementing BGP, yes. If there is a library out there we can use it.
 
> Another suggestion is to unify the p2p connection management (candidates, offers, etc) with this routing protocol, and move that logic to this RoutingService. That way, the server can have full context of the mesh at all times (removing the need of reporting) and optimize the network accordingly. There is more: implementing vanilla libwebrtc based comms in native clients will be trivial, since the logic to pick the candidates will be entirely server-based. Greatly benefiting "the protocol" and the implementer's lives. Along with consistent fixes applied to servers without needing to redistribute the native clients to all devices and across all app-stores.

As stated above the messaing service is mostly a fallback, if the network has no clusters and it's stable, the service won't be used.

Reply:
We can continue this in another PR, I'm resolving the conversation in order to merge this one.